diff --git a/src/hotspot/cpu/riscv/c2_MacroAssembler_riscv.cpp b/src/hotspot/cpu/riscv/c2_MacroAssembler_riscv.cpp
index 2051bd657..5615b26c0 100644
--- a/src/hotspot/cpu/riscv/c2_MacroAssembler_riscv.cpp
+++ b/src/hotspot/cpu/riscv/c2_MacroAssembler_riscv.cpp
@@ -2477,8 +2477,11 @@ void C2_MacroAssembler::float_to_float16_v(VectorRegister dst, VectorRegister sr
   bind(stub->continuation());
 }
 
-void C2_MacroAssembler::signum_fp_v(VectorRegister dst, VectorRegister one, BasicType bt, int vlen) {
-  vsetvli_helper(bt, vlen, Assembler::mu);
+void C2_MacroAssembler::signum_fp_v(VectorRegister dst, VectorRegister one, BasicType bt, int vlen,
+                                    bool set_vtype) {
+  if (set_vtype) {
+    vsetvli_helper(bt, vlen, Assembler::mu);
+  }
 
   // check if input is -0, +0, signaling NaN or quiet NaN
   vfclass_v(v0, dst);
@@ -2497,7 +2500,7 @@ void C2_MacroAssembler::signum_fp_v(VectorRegister dst, VectorRegister one, Basi
 //    float >= Integer.MAX_VALUE,
 //    float <= Integer.MIN_VALUE.
 void C2_MacroAssembler::java_round_float_v(VectorRegister dst, VectorRegister src, FloatRegister ftmp,
-                                           BasicType bt, uint vector_length) {
+                                           BasicType bt, uint vector_length, bool set_vtype) {
   // In riscv, there is no straight corresponding rounding mode to satisfy the behaviour defined,
   // in java api spec, i.e. any rounding mode can not handle some corner cases, e.g.
   //  RNE is the closest one, but it ties to "even", which means 1.5/2.5 both will be converted
@@ -2515,7 +2518,9 @@ void C2_MacroAssembler::java_round_float_v(VectorRegister dst, VectorRegister sr
   // Check MacroAssembler::java_round_float and C2_MacroAssembler::vector_round_sve in aarch64 for more details.
 
   csrwi(CSR_FRM, C2_MacroAssembler::rdn);
-  vsetvli_helper(bt, vector_length, Assembler::mu);
+  if (set_vtype) {
+    vsetvli_helper(bt, vector_length, Assembler::mu);
+  }
 
   // don't rearrage the instructions sequence order without performance testing.
   // check MacroAssembler::java_round_float in riscv64 for more details.
@@ -2537,11 +2542,13 @@ void C2_MacroAssembler::java_round_float_v(VectorRegister dst, VectorRegister sr
 // java.lang.Math.round(double a)
 // Returns the closest long to the argument, with ties rounding to positive infinity.
 void C2_MacroAssembler::java_round_double_v(VectorRegister dst, VectorRegister src, FloatRegister ftmp,
-                                            BasicType bt, uint vector_length) {
+                                            BasicType bt, uint vector_length, bool set_vtype) {
   // check C2_MacroAssembler::java_round_float_v above for more details.
 
   csrwi(CSR_FRM, C2_MacroAssembler::rdn);
-  vsetvli_helper(bt, vector_length, Assembler::mu);
+  if (set_vtype) {
+    vsetvli_helper(bt, vector_length, Assembler::mu);
+  }
 
   mv(t0, julong_cast(0.5));
   fmv_d_x(ftmp, t0);
@@ -2870,10 +2877,12 @@ void C2_MacroAssembler::string_indexof_char_v(Register str1, Register cnt1,
 
 // Set dst to NaN if any NaN input.
 void C2_MacroAssembler::minmax_fp_v(VectorRegister dst, VectorRegister src1, VectorRegister src2,
-                                    BasicType bt, bool is_min, uint vector_length) {
+                                    BasicType bt, bool is_min, uint vector_length, bool set_vtype) {
   assert_different_registers(dst, src1, src2);
 
-  vsetvli_helper(bt, vector_length, Assembler::mu);
+  if (set_vtype) {
+    vsetvli_helper(bt, vector_length, Assembler::mu);
+  }
 
   is_min ? vfmin_vv(dst, src1, src2)
          : vfmax_vv(dst, src1, src2);
@@ -2889,9 +2898,11 @@ void C2_MacroAssembler::minmax_fp_v(VectorRegister dst, VectorRegister src1, Vec
 // are handled with a mask-undisturbed policy.
 void C2_MacroAssembler::minmax_fp_masked_v(VectorRegister dst, VectorRegister src1, VectorRegister src2,
                                            VectorRegister vmask, VectorRegister tmp1, VectorRegister tmp2,
-                                           BasicType bt, bool is_min, uint vector_length) {
+                                           BasicType bt, bool is_min, uint vector_length, bool set_vtype) {
   assert_different_registers(src1, src2, tmp1, tmp2);
-  vsetvli_helper(bt, vector_length, Assembler::mu);
+  if (set_vtype) {
+    vsetvli_helper(bt, vector_length, Assembler::mu);
+  }
 
   // Check vector elements of src1 and src2 for NaN.
   vmfeq_vv(tmp1, src1, src1);
@@ -2912,7 +2923,8 @@ void C2_MacroAssembler::minmax_fp_masked_v(VectorRegister dst, VectorRegister sr
 void C2_MacroAssembler::reduce_minmax_fp_v(FloatRegister dst,
                                            FloatRegister src1, VectorRegister src2,
                                            VectorRegister tmp1, VectorRegister tmp2,
-                                           bool is_double, bool is_min, uint vector_length, VectorMask vm) {
+                                           bool is_double, bool is_min, uint vector_length, VectorMask vm,
+                                           bool set_vtype) {
   assert_different_registers(dst, src1);
   assert_different_registers(src2, tmp1, tmp2);
 
@@ -2922,7 +2934,9 @@ void C2_MacroAssembler::reduce_minmax_fp_v(FloatRegister dst,
             : feq_s(t0, src1, src1);
   beqz(t0, L_NaN_2);
 
-  vsetvli_helper(is_double ? T_DOUBLE : T_FLOAT, vector_length, vm == Assembler::v0_t ? Assembler::mu : Assembler::ma);
+  if (set_vtype) {
+    vsetvli_helper(is_double ? T_DOUBLE : T_FLOAT, vector_length, vm == Assembler::v0_t ? Assembler::mu : Assembler::ma);
+  }
   vfmv_s_f(tmp2, src1);
 
   is_min ? vfredmin_vs(tmp1, src2, tmp2, vm)
@@ -2957,9 +2971,12 @@ bool C2_MacroAssembler::in_scratch_emit_size() {
 
 void C2_MacroAssembler::reduce_integral_v(Register dst, Register src1,
                                           VectorRegister src2, VectorRegister tmp,
-                                          int opc, BasicType bt, uint vector_length, VectorMask vm) {
+                                          int opc, BasicType bt, uint vector_length, VectorMask vm,
+                                          bool set_vtype) {
   assert(bt == T_BYTE || bt == T_SHORT || bt == T_INT || bt == T_LONG, "unsupported element type");
-  vsetvli_helper(bt, vector_length, vm == Assembler::v0_t ? Assembler::mu : Assembler::ma);
+  if (set_vtype) {
+    vsetvli_helper(bt, vector_length, vm == Assembler::v0_t ? Assembler::mu : Assembler::ma);
+  }
   vmv_s_x(tmp, src1);
   switch (opc) {
     case Op_AddReductionVI:
@@ -3041,10 +3058,13 @@ void C2_MacroAssembler::vsetvli_helper(BasicType bt, uint vector_length, LMUL vl
 }
 
 void C2_MacroAssembler::compare_integral_v(VectorRegister vd, VectorRegister src1, VectorRegister src2,
-                                           int cond, BasicType bt, uint vector_length, VectorMask vm) {
+                                           int cond, BasicType bt, uint vector_length, VectorMask vm,
+                                           bool set_vtype) {
   assert(is_integral_type(bt), "unsupported element type");
   assert(vm == Assembler::v0_t ? vd != v0 : true, "should be different registers");
-  vsetvli_helper(bt, vector_length, vm == Assembler::v0_t ? Assembler::mu : Assembler::ma);
+  if (set_vtype) {
+    vsetvli_helper(bt, vector_length, vm == Assembler::v0_t ? Assembler::mu : Assembler::ma);
+  }
   if (vm == Assembler::v0_t) {
     vmclr_m(vd);
   }
@@ -3066,10 +3086,13 @@ void C2_MacroAssembler::compare_integral_v(VectorRegister vd, VectorRegister src
 }
 
 void C2_MacroAssembler::compare_fp_v(VectorRegister vd, VectorRegister src1, VectorRegister src2,
-                                     int cond, BasicType bt, uint vector_length, VectorMask vm) {
+                                     int cond, BasicType bt, uint vector_length, VectorMask vm,
+                                     bool set_vtype) {
   assert(is_floating_point_type(bt), "unsupported element type");
   assert(vm == Assembler::v0_t ? vd != v0 : true, "should be different registers");
-  vsetvli_helper(bt, vector_length, vm == Assembler::v0_t ? Assembler::mu : Assembler::ma);
+  if (set_vtype) {
+    vsetvli_helper(bt, vector_length, vm == Assembler::v0_t ? Assembler::mu : Assembler::ma);
+  }
   if (vm == Assembler::v0_t) {
     vmclr_m(vd);
   }
@@ -3102,7 +3125,8 @@ void C2_MacroAssembler::unspill_vmask(VectorRegister v, int offset) {
 }
 
 void C2_MacroAssembler::integer_extend_v(VectorRegister dst, BasicType dst_bt, uint vector_length,
-                                         VectorRegister src, BasicType src_bt, bool is_signed) {
+                                         VectorRegister src, BasicType src_bt, bool is_signed,
+                                         bool set_vtype) {
   assert(type2aelembytes(dst_bt) > type2aelembytes(src_bt) && type2aelembytes(dst_bt) <= 8 && type2aelembytes(src_bt) <= 4, "invalid element size");
   assert(dst_bt != T_FLOAT && dst_bt != T_DOUBLE && src_bt != T_FLOAT && src_bt != T_DOUBLE, "unsupported element type");
   // https://github.com/riscv/riscv-v-spec/blob/master/v-spec.adoc#52-vector-operands
@@ -3111,7 +3135,9 @@ void C2_MacroAssembler::integer_extend_v(VectorRegister dst, BasicType dst_bt, u
   // Since LMUL=1, vd and vs cannot be the same.
   assert_different_registers(dst, src);
 
-  vsetvli_helper(dst_bt, vector_length);
+  if (set_vtype) {
+    vsetvli_helper(dst_bt, vector_length);
+  }
   if (is_signed) {
     if (src_bt == T_BYTE) {
       switch (dst_bt) {
diff --git a/src/hotspot/cpu/riscv/c2_MacroAssembler_riscv.hpp b/src/hotspot/cpu/riscv/c2_MacroAssembler_riscv.hpp
index 07d4fc499..8a3e7ad85 100644
--- a/src/hotspot/cpu/riscv/c2_MacroAssembler_riscv.hpp
+++ b/src/hotspot/cpu/riscv/c2_MacroAssembler_riscv.hpp
@@ -195,13 +195,16 @@
   void float16_to_float(FloatRegister dst, Register src, Register tmp);
   void float_to_float16(Register dst, FloatRegister src, FloatRegister ftmp, Register xtmp);
 
-  void signum_fp_v(VectorRegister dst, VectorRegister one, BasicType bt, int vlen);
+  void signum_fp_v(VectorRegister dst, VectorRegister one, BasicType bt, int vlen,
+                   bool set_vtype = true);
 
 
   // intrinsic methods implemented by rvv instructions
 
-  void java_round_float_v(VectorRegister dst, VectorRegister src, FloatRegister ftmp, BasicType bt, uint vector_length);
-  void java_round_double_v(VectorRegister dst, VectorRegister src, FloatRegister ftmp, BasicType bt, uint vector_length);
+  void java_round_float_v(VectorRegister dst, VectorRegister src, FloatRegister ftmp,
+                          BasicType bt, uint vector_length, bool set_vtype = true);
+  void java_round_double_v(VectorRegister dst, VectorRegister src, FloatRegister ftmp,
+                           BasicType bt, uint vector_length, bool set_vtype = true);
 
   void float16_to_float_v(VectorRegister dst, VectorRegister src, uint vector_length);
   void float_to_float16_v(VectorRegister dst, VectorRegister src, VectorRegister vtmp, Register tmp, uint vector_length);
@@ -242,22 +245,26 @@
 
   void minmax_fp_v(VectorRegister dst,
                   VectorRegister src1, VectorRegister src2,
-                  BasicType bt, bool is_min, uint vector_length);
+                  BasicType bt, bool is_min, uint vector_length,
+                  bool set_vtype = true);
 
   void minmax_fp_masked_v(VectorRegister dst, VectorRegister src1, VectorRegister src2,
                           VectorRegister vmask, VectorRegister tmp1, VectorRegister tmp2,
-                          BasicType bt, bool is_min, uint vector_length);
+                          BasicType bt, bool is_min, uint vector_length,
+                          bool set_vtype = true);
 
   void reduce_minmax_fp_v(FloatRegister dst,
                           FloatRegister src1, VectorRegister src2,
                           VectorRegister tmp1, VectorRegister tmp2,
                           bool is_double, bool is_min, uint vector_length,
-                          VectorMask vm = Assembler::unmasked);
+                          VectorMask vm = Assembler::unmasked,
+                          bool set_vtype = true);
 
   void reduce_integral_v(Register dst, Register src1,
                         VectorRegister src2, VectorRegister tmp,
                         int opc, BasicType bt, uint vector_length,
-                        VectorMask vm = Assembler::unmasked);
+                        VectorMask vm = Assembler::unmasked,
+                        bool set_vtype = true);
 
   void reduce_mul_integral_v(Register dst, Register src1, VectorRegister src2,
                              VectorRegister vtmp1, VectorRegister vtmp2, BasicType bt,
@@ -271,10 +278,12 @@
   }
 
   void compare_integral_v(VectorRegister dst, VectorRegister src1, VectorRegister src2, int cond,
-                          BasicType bt, uint vector_length, VectorMask vm = Assembler::unmasked);
+                          BasicType bt, uint vector_length, VectorMask vm = Assembler::unmasked,
+                          bool set_vtype = true);
 
   void compare_fp_v(VectorRegister dst, VectorRegister src1, VectorRegister src2, int cond,
-                    BasicType bt, uint vector_length, VectorMask vm = Assembler::unmasked);
+                    BasicType bt, uint vector_length, VectorMask vm = Assembler::unmasked,
+                    bool set_vtype = true);
 
   void spill_vmask(VectorRegister v, int offset);
 
@@ -289,7 +298,8 @@
   }
 
   void integer_extend_v(VectorRegister dst, BasicType dst_bt, uint vector_length,
-                        VectorRegister src, BasicType src_bt, bool is_signed);
+                        VectorRegister src, BasicType src_bt, bool is_signed,
+                        bool set_vtype = true);
 
   void integer_narrow_v(VectorRegister dst, BasicType dst_bt, uint vector_length,
                         VectorRegister src, BasicType src_bt);
diff --git a/src/hotspot/cpu/riscv/riscv_v.ad b/src/hotspot/cpu/riscv/riscv_v.ad
index 85ef8046a..7772d6960 100644
--- a/src/hotspot/cpu/riscv/riscv_v.ad
+++ b/src/hotspot/cpu/riscv/riscv_v.ad
@@ -246,14 +246,15 @@ instruct vmaskcmp(vRegMask dst, vReg src1, vReg src2, immI cond) %{
             Matcher::vector_element_basic_type(n) == T_INT ||
             Matcher::vector_element_basic_type(n) == T_LONG);
   match(Set dst (VectorMaskCmp (Binary src1 src2) cond));
-  ins_riscv_vset(false);
+  ins_riscv_vset(riscv_vset_fixed(Matcher::vector_element_basic_type(this), Matcher::vector_length(this), req, Assembler::m1, Assembler::ma));
   format %{ "vmaskcmp $dst, $src1, $src2, $cond" %}
   ins_encode %{
     BasicType bt = Matcher::vector_element_basic_type(this);
     uint vector_length = Matcher::vector_length(this);
     __ compare_integral_v(as_VectorRegister($dst$$reg),
                           as_VectorRegister($src1$$reg), as_VectorRegister($src2$$reg),
-                          (int)($cond$$constant), bt, vector_length);
+                          (int)($cond$$constant), bt, vector_length, Assembler::unmasked,
+                          false /* set_vtype */);
   %}
   ins_pipe(pipe_slow);
 %}
@@ -265,14 +266,15 @@ instruct vmaskcmp_masked(vRegMask dst, vReg src1, vReg src2, immI cond, vRegMask
             Matcher::vector_element_basic_type(n) == T_LONG);
   match(Set dst (VectorMaskCmp (Binary src1 src2) (Binary cond v0)));
   effect(TEMP_DEF dst);
-  ins_riscv_vset(false);
+  ins_riscv_vset(riscv_vset_fixed(Matcher::vector_element_basic_type(this), Matcher::vector_length(this), req, Assembler::m1, Assembler::mu));
   format %{ "vmaskcmp_masked $dst, $src1, $src2, $cond, $v0" %}
   ins_encode %{
     BasicType bt = Matcher::vector_element_basic_type(this);
     uint vector_length = Matcher::vector_length(this);
     __ compare_integral_v(as_VectorRegister($dst$$reg),
                           as_VectorRegister($src1$$reg), as_VectorRegister($src2$$reg),
-                          (int)($cond$$constant), bt, vector_length, Assembler::v0_t);
+                          (int)($cond$$constant), bt, vector_length, Assembler::v0_t,
+                          false /* set_vtype */);
   %}
   ins_pipe(pipe_slow);
 %}
@@ -283,14 +285,15 @@ instruct vmaskcmp_fp(vRegMask dst, vReg src1, vReg src2, immI cond) %{
   predicate(Matcher::vector_element_basic_type(n) == T_FLOAT ||
             Matcher::vector_element_basic_type(n) == T_DOUBLE);
   match(Set dst (VectorMaskCmp (Binary src1 src2) cond));
-  ins_riscv_vset(false);
+  ins_riscv_vset(riscv_vset_fixed(Matcher::vector_element_basic_type(this), Matcher::vector_length(this), req, Assembler::m1, Assembler::ma));
   format %{ "vmaskcmp_fp $dst, $src1, $src2, $cond" %}
   ins_encode %{
     BasicType bt = Matcher::vector_element_basic_type(this);
     uint vector_length = Matcher::vector_length(this);
     __ compare_fp_v(as_VectorRegister($dst$$reg),
                     as_VectorRegister($src1$$reg), as_VectorRegister($src2$$reg),
-                    (int)($cond$$constant), bt, vector_length);
+                    (int)($cond$$constant), bt, vector_length, Assembler::unmasked,
+                          false /* set_vtype */);
   %}
   ins_pipe(pipe_slow);
 %}
@@ -300,14 +303,15 @@ instruct vmaskcmp_fp_masked(vRegMask dst, vReg src1, vReg src2, immI cond, vRegM
             Matcher::vector_element_basic_type(n) == T_DOUBLE);
   match(Set dst (VectorMaskCmp (Binary src1 src2) (Binary cond v0)));
   effect(TEMP_DEF dst);
-  ins_riscv_vset(false);
+  ins_riscv_vset(riscv_vset_fixed(Matcher::vector_element_basic_type(this), Matcher::vector_length(this), req, Assembler::m1, Assembler::mu));
   format %{ "vmaskcmp_fp_masked $dst, $src1, $src2, $cond, $v0" %}
   ins_encode %{
     BasicType bt = Matcher::vector_element_basic_type(this);
     uint vector_length = Matcher::vector_length(this);
     __ compare_fp_v(as_VectorRegister($dst$$reg),
                     as_VectorRegister($src1$$reg), as_VectorRegister($src2$$reg),
-                    (int)($cond$$constant), bt, vector_length, Assembler::v0_t);
+                    (int)($cond$$constant), bt, vector_length, Assembler::v0_t,
+                          false /* set_vtype */);
   %}
   ins_pipe(pipe_slow);
 %}
@@ -1698,14 +1702,15 @@ instruct vminu_masked(vReg dst_src1, vReg src2, vRegMask_V0 v0) %{
 instruct vmax_hfp(vReg dst, vReg src1, vReg src2, vRegMask_V0 v0) %{
   match(Set dst (MaxVHF src1 src2));
   effect(TEMP_DEF dst, TEMP v0);
-  ins_riscv_vset(false);
+  ins_riscv_vset(riscv_vset_fixed(T_SHORT, Matcher::vector_length(this), req, Assembler::m1, Assembler::mu));
   format %{ "vmax_hfp $dst, $src1, $src2" %}
   ins_encode %{
     assert(UseZvfh, "must");
     assert(Matcher::vector_element_basic_type(this) == T_SHORT, "must");
     __ minmax_fp_v(as_VectorRegister($dst$$reg),
                    as_VectorRegister($src1$$reg), as_VectorRegister($src2$$reg),
-                   T_SHORT, false /* is_min */, Matcher::vector_length(this));
+                   T_SHORT, false /* is_min */, Matcher::vector_length(this),
+                   false /* set_vtype */);
   %}
   ins_pipe(pipe_slow);
 %}
@@ -1713,14 +1718,15 @@ instruct vmax_hfp(vReg dst, vReg src1, vReg src2, vRegMask_V0 v0) %{
 instruct vmin_hfp(vReg dst, vReg src1, vReg src2, vRegMask_V0 v0) %{
   match(Set dst (MinVHF src1 src2));
   effect(TEMP_DEF dst, TEMP v0);
-  ins_riscv_vset(false);
+  ins_riscv_vset(riscv_vset_fixed(T_SHORT, Matcher::vector_length(this), req, Assembler::m1, Assembler::mu));
   format %{ "vmin_hfp $dst, $src1, $src2" %}
   ins_encode %{
     assert(UseZvfh, "must");
     assert(Matcher::vector_element_basic_type(this) == T_SHORT, "must");
     __ minmax_fp_v(as_VectorRegister($dst$$reg),
                    as_VectorRegister($src1$$reg), as_VectorRegister($src2$$reg),
-                   T_SHORT, true /* is_min */, Matcher::vector_length(this));
+                   T_SHORT, true /* is_min */, Matcher::vector_length(this),
+                   false /* set_vtype */);
   %}
   ins_pipe(pipe_slow);
 %}
@@ -1732,13 +1738,14 @@ instruct vmax_fp(vReg dst, vReg src1, vReg src2, vRegMask_V0 v0) %{
             Matcher::vector_element_basic_type(n) == T_DOUBLE);
   match(Set dst (MaxV src1 src2));
   effect(TEMP_DEF dst, TEMP v0);
-  ins_riscv_vset(false);
+  ins_riscv_vset(riscv_vset_from_vect(this, req, Assembler::m1, Assembler::mu));
   format %{ "vmax_fp $dst, $src1, $src2" %}
   ins_encode %{
     BasicType bt = Matcher::vector_element_basic_type(this);
     __ minmax_fp_v(as_VectorRegister($dst$$reg),
                    as_VectorRegister($src1$$reg), as_VectorRegister($src2$$reg),
-                   bt, false /* is_min */, Matcher::vector_length(this));
+                   bt, false /* is_min */, Matcher::vector_length(this),
+                   false /* set_vtype */);
   %}
   ins_pipe(pipe_slow);
 %}
@@ -1748,13 +1755,14 @@ instruct vmin_fp(vReg dst, vReg src1, vReg src2, vRegMask_V0 v0) %{
             Matcher::vector_element_basic_type(n) == T_DOUBLE);
   match(Set dst (MinV src1 src2));
   effect(TEMP_DEF dst, TEMP v0);
-  ins_riscv_vset(false);
+  ins_riscv_vset(riscv_vset_from_vect(this, req, Assembler::m1, Assembler::mu));
   format %{ "vmin_fp $dst, $src1, $src2" %}
   ins_encode %{
     BasicType bt = Matcher::vector_element_basic_type(this);
     __ minmax_fp_v(as_VectorRegister($dst$$reg),
                    as_VectorRegister($src1$$reg), as_VectorRegister($src2$$reg),
-                   bt, true /* is_min */, Matcher::vector_length(this));
+                   bt, true /* is_min */, Matcher::vector_length(this),
+                   false /* set_vtype */);
   %}
   ins_pipe(pipe_slow);
 %}
@@ -1766,14 +1774,15 @@ instruct vmax_fp_masked(vReg dst_src1, vReg src2, vRegMask vmask, vReg tmp1, vRe
             Matcher::vector_element_basic_type(n) == T_DOUBLE);
   match(Set dst_src1 (MaxV (Binary dst_src1 src2) vmask));
   effect(TEMP_DEF dst_src1, TEMP tmp1, TEMP tmp2, TEMP v0);
-  ins_riscv_vset(false);
+  ins_riscv_vset(riscv_vset_from_vect(this, req, Assembler::m1, Assembler::mu));
   format %{ "vmax_fp_masked $dst_src1, $dst_src1, $src2, $vmask\t# KILL $tmp1, $tmp2, $v0" %}
   ins_encode %{
     BasicType bt = Matcher::vector_element_basic_type(this);
     __ minmax_fp_masked_v(as_VectorRegister($dst_src1$$reg), as_VectorRegister($dst_src1$$reg),
                           as_VectorRegister($src2$$reg), as_VectorRegister($vmask$$reg),
                           as_VectorRegister($tmp1$$reg), as_VectorRegister($tmp2$$reg),
-                          bt, false /* is_min */, Matcher::vector_length(this));
+                          bt, false /* is_min */, Matcher::vector_length(this),
+                          false /* set_vtype */);
   %}
   ins_pipe(pipe_slow);
 %}
@@ -1783,14 +1792,15 @@ instruct vmin_fp_masked(vReg dst_src1, vReg src2, vRegMask vmask, vReg tmp1, vRe
             Matcher::vector_element_basic_type(n) == T_DOUBLE);
   match(Set dst_src1 (MinV (Binary dst_src1 src2) vmask));
   effect(TEMP_DEF dst_src1, TEMP tmp1, TEMP tmp2, TEMP v0);
-  ins_riscv_vset(false);
+  ins_riscv_vset(riscv_vset_from_vect(this, req, Assembler::m1, Assembler::mu));
   format %{ "vmin_fp_masked $dst_src1, $dst_src1, $src2, $vmask\t# KILL $tmp1, $tmp2, $v0" %}
   ins_encode %{
     BasicType bt = Matcher::vector_element_basic_type(this);
     __ minmax_fp_masked_v(as_VectorRegister($dst_src1$$reg), as_VectorRegister($dst_src1$$reg),
                           as_VectorRegister($src2$$reg), as_VectorRegister($vmask$$reg),
                           as_VectorRegister($tmp1$$reg), as_VectorRegister($tmp2$$reg),
-                          bt, true /* is_min */, Matcher::vector_length(this));
+                          bt, true /* is_min */, Matcher::vector_length(this),
+                          false /* set_vtype */);
   %}
   ins_pipe(pipe_slow);
 %}
@@ -2231,13 +2241,14 @@ instruct reduce_and(iRegINoSp dst, iRegIorL2I src1, vReg src2, vReg tmp) %{
             Matcher::vector_element_basic_type(n->in(2)) == T_INT);
   match(Set dst (AndReductionV src1 src2));
   effect(TEMP tmp);
-  ins_riscv_vset(false);
+  ins_riscv_vset(riscv_vset_from_node(this, req, Assembler::m1, Assembler::ma));
   format %{ "reduce_and $dst, $src1, $src2\t# KILL $tmp" %}
   ins_encode %{
     BasicType bt = Matcher::vector_element_basic_type(this, $src2);
     __ reduce_integral_v($dst$$Register, $src1$$Register,
                          as_VectorRegister($src2$$reg), as_VectorRegister($tmp$$reg),
-                         this->ideal_Opcode(), bt, Matcher::vector_length(this, $src2));
+                         this->ideal_Opcode(), bt, Matcher::vector_length(this, $src2),
+                         Assembler::unmasked, false /* set_vtype */);
   %}
   ins_pipe(pipe_slow);
 %}
@@ -2246,13 +2257,14 @@ instruct reduce_andL(iRegLNoSp dst, iRegL src1, vReg src2, vReg tmp) %{
   predicate(Matcher::vector_element_basic_type(n->in(2)) == T_LONG);
   match(Set dst (AndReductionV src1 src2));
   effect(TEMP tmp);
-  ins_riscv_vset(false);
+  ins_riscv_vset(riscv_vset_from_node(this, req, Assembler::m1, Assembler::ma));
   format %{ "reduce_andL $dst, $src1, $src2\t# KILL $tmp" %}
   ins_encode %{
     BasicType bt = Matcher::vector_element_basic_type(this, $src2);
     __ reduce_integral_v($dst$$Register, $src1$$Register,
                          as_VectorRegister($src2$$reg), as_VectorRegister($tmp$$reg),
-                         this->ideal_Opcode(), bt, Matcher::vector_length(this, $src2));
+                         this->ideal_Opcode(), bt, Matcher::vector_length(this, $src2),
+                         Assembler::unmasked, false /* set_vtype */);
   %}
   ins_pipe(pipe_slow);
 %}
@@ -2265,14 +2277,14 @@ instruct reduce_and_masked(iRegINoSp dst, iRegIorL2I src1, vReg src2, vRegMask_V
             Matcher::vector_element_basic_type(n->in(2)) == T_INT);
   match(Set dst (AndReductionV (Binary src1 src2) v0));
   effect(TEMP tmp);
-  ins_riscv_vset(false);
+  ins_riscv_vset(riscv_vset_from_node(this, req, Assembler::m1, Assembler::mu));
   format %{ "reduce_and_masked $dst, $src1, $src2, $v0\t# KILL $tmp" %}
   ins_encode %{
     BasicType bt = Matcher::vector_element_basic_type(this, $src2);
     __ reduce_integral_v($dst$$Register, $src1$$Register,
                          as_VectorRegister($src2$$reg), as_VectorRegister($tmp$$reg),
                          this->ideal_Opcode(), bt, Matcher::vector_length(this, $src2),
-                         Assembler::v0_t);
+                         Assembler::v0_t, false /* set_vtype */);
   %}
   ins_pipe(pipe_slow);
 %}
@@ -2281,14 +2293,14 @@ instruct reduce_andL_masked(iRegLNoSp dst, iRegL src1, vReg src2, vRegMask_V0 v0
   predicate(Matcher::vector_element_basic_type(n->in(2)) == T_LONG);
   match(Set dst (AndReductionV (Binary src1 src2) v0));
   effect(TEMP tmp);
-  ins_riscv_vset(false);
+  ins_riscv_vset(riscv_vset_from_node(this, req, Assembler::m1, Assembler::mu));
   format %{ "reduce_andL_masked $dst, $src1, $src2, $v0\t# KILL $tmp" %}
   ins_encode %{
     BasicType bt = Matcher::vector_element_basic_type(this, $src2);
     __ reduce_integral_v($dst$$Register, $src1$$Register,
                          as_VectorRegister($src2$$reg), as_VectorRegister($tmp$$reg),
                          this->ideal_Opcode(), bt, Matcher::vector_length(this, $src2),
-                         Assembler::v0_t);
+                         Assembler::v0_t, false /* set_vtype */);
   %}
   ins_pipe(pipe_slow);
 %}
@@ -2301,13 +2313,14 @@ instruct reduce_or(iRegINoSp dst, iRegIorL2I src1, vReg src2, vReg tmp) %{
             Matcher::vector_element_basic_type(n->in(2)) == T_INT);
   match(Set dst (OrReductionV src1 src2));
   effect(TEMP tmp);
-  ins_riscv_vset(false);
+  ins_riscv_vset(riscv_vset_from_node(this, req, Assembler::m1, Assembler::ma));
   format %{ "reduce_or $dst, $src1, $src2\t# KILL $tmp" %}
   ins_encode %{
     BasicType bt = Matcher::vector_element_basic_type(this, $src2);
     __ reduce_integral_v($dst$$Register, $src1$$Register,
                          as_VectorRegister($src2$$reg), as_VectorRegister($tmp$$reg),
-                         this->ideal_Opcode(), bt, Matcher::vector_length(this, $src2));
+                         this->ideal_Opcode(), bt, Matcher::vector_length(this, $src2),
+                         Assembler::unmasked, false /* set_vtype */);
   %}
   ins_pipe(pipe_slow);
 %}
@@ -2316,13 +2329,14 @@ instruct reduce_orL(iRegLNoSp dst, iRegL src1, vReg src2, vReg tmp) %{
   predicate(Matcher::vector_element_basic_type(n->in(2)) == T_LONG);
   match(Set dst (OrReductionV src1 src2));
   effect(TEMP tmp);
-  ins_riscv_vset(false);
+  ins_riscv_vset(riscv_vset_from_node(this, req, Assembler::m1, Assembler::ma));
   format %{ "reduce_orL $dst, $src1, $src2\t# KILL $tmp" %}
   ins_encode %{
     BasicType bt = Matcher::vector_element_basic_type(this, $src2);
     __ reduce_integral_v($dst$$Register, $src1$$Register,
                          as_VectorRegister($src2$$reg), as_VectorRegister($tmp$$reg),
-                         this->ideal_Opcode(), bt, Matcher::vector_length(this, $src2));
+                         this->ideal_Opcode(), bt, Matcher::vector_length(this, $src2),
+                         Assembler::unmasked, false /* set_vtype */);
   %}
   ins_pipe(pipe_slow);
 %}
@@ -2335,14 +2349,14 @@ instruct reduce_or_masked(iRegINoSp dst, iRegIorL2I src1, vReg src2, vRegMask_V0
             Matcher::vector_element_basic_type(n->in(2)) == T_INT);
   match(Set dst (OrReductionV (Binary src1 src2) v0));
   effect(TEMP tmp);
-  ins_riscv_vset(false);
+  ins_riscv_vset(riscv_vset_from_node(this, req, Assembler::m1, Assembler::mu));
   format %{ "reduce_or_masked $dst, $src1, $src2, $v0\t# KILL $tmp" %}
   ins_encode %{
     BasicType bt = Matcher::vector_element_basic_type(this, $src2);
     __ reduce_integral_v($dst$$Register, $src1$$Register,
                          as_VectorRegister($src2$$reg), as_VectorRegister($tmp$$reg),
                          this->ideal_Opcode(), bt, Matcher::vector_length(this, $src2),
-                         Assembler::v0_t);
+                         Assembler::v0_t, false /* set_vtype */);
   %}
   ins_pipe(pipe_slow);
 %}
@@ -2351,14 +2365,14 @@ instruct reduce_orL_masked(iRegLNoSp dst, iRegL src1, vReg src2, vRegMask_V0 v0,
   predicate(Matcher::vector_element_basic_type(n->in(2)) == T_LONG);
   match(Set dst (OrReductionV (Binary src1 src2) v0));
   effect(TEMP tmp);
-  ins_riscv_vset(false);
+  ins_riscv_vset(riscv_vset_from_node(this, req, Assembler::m1, Assembler::mu));
   format %{ "reduce_orL_masked $dst, $src1, $src2, $v0\t# KILL $tmp" %}
   ins_encode %{
     BasicType bt = Matcher::vector_element_basic_type(this, $src2);
     __ reduce_integral_v($dst$$Register, $src1$$Register,
                          as_VectorRegister($src2$$reg), as_VectorRegister($tmp$$reg),
                          this->ideal_Opcode(), bt, Matcher::vector_length(this, $src2),
-                         Assembler::v0_t);
+                         Assembler::v0_t, false /* set_vtype */);
   %}
   ins_pipe(pipe_slow);
 %}
@@ -2371,13 +2385,14 @@ instruct reduce_xor(iRegINoSp dst, iRegIorL2I src1, vReg src2, vReg tmp) %{
             Matcher::vector_element_basic_type(n->in(2)) == T_INT);
   match(Set dst (XorReductionV src1 src2));
   effect(TEMP tmp);
-  ins_riscv_vset(false);
+  ins_riscv_vset(riscv_vset_from_node(this, req, Assembler::m1, Assembler::ma));
   format %{ "reduce_xor $dst, $src1, $src2\t# KILL $tmp" %}
   ins_encode %{
     BasicType bt = Matcher::vector_element_basic_type(this, $src2);
     __ reduce_integral_v($dst$$Register, $src1$$Register,
                          as_VectorRegister($src2$$reg), as_VectorRegister($tmp$$reg),
-                         this->ideal_Opcode(), bt, Matcher::vector_length(this, $src2));
+                         this->ideal_Opcode(), bt, Matcher::vector_length(this, $src2),
+                         Assembler::unmasked, false /* set_vtype */);
   %}
   ins_pipe(pipe_slow);
 %}
@@ -2386,13 +2401,14 @@ instruct reduce_xorL(iRegLNoSp dst, iRegL src1, vReg src2, vReg tmp) %{
   predicate(Matcher::vector_element_basic_type(n->in(2)) == T_LONG);
   match(Set dst (XorReductionV src1 src2));
   effect(TEMP tmp);
-  ins_riscv_vset(false);
+  ins_riscv_vset(riscv_vset_from_node(this, req, Assembler::m1, Assembler::ma));
   format %{ "reduce_xorL $dst, $src1, $src2\t# KILL $tmp" %}
   ins_encode %{
     BasicType bt = Matcher::vector_element_basic_type(this, $src2);
     __ reduce_integral_v($dst$$Register, $src1$$Register,
                          as_VectorRegister($src2$$reg), as_VectorRegister($tmp$$reg),
-                         this->ideal_Opcode(), bt, Matcher::vector_length(this, $src2));
+                         this->ideal_Opcode(), bt, Matcher::vector_length(this, $src2),
+                         Assembler::unmasked, false /* set_vtype */);
   %}
   ins_pipe(pipe_slow);
 %}
@@ -2405,14 +2421,14 @@ instruct reduce_xor_masked(iRegINoSp dst, iRegIorL2I src1, vReg src2, vRegMask_V
             Matcher::vector_element_basic_type(n->in(2)) == T_INT);
   match(Set dst (XorReductionV (Binary src1 src2) v0));
   effect(TEMP tmp);
-  ins_riscv_vset(false);
+  ins_riscv_vset(riscv_vset_from_node(this, req, Assembler::m1, Assembler::mu));
   format %{ "reduce_xor_masked $dst, $src1, $src2, $v0\t# KILL $tmp" %}
   ins_encode %{
     BasicType bt = Matcher::vector_element_basic_type(this, $src2);
     __ reduce_integral_v($dst$$Register, $src1$$Register,
                          as_VectorRegister($src2$$reg), as_VectorRegister($tmp$$reg),
                          this->ideal_Opcode(), bt, Matcher::vector_length(this, $src2),
-                         Assembler::v0_t);
+                         Assembler::v0_t, false /* set_vtype */);
   %}
   ins_pipe(pipe_slow);
 %}
@@ -2421,14 +2437,14 @@ instruct reduce_xorL_masked(iRegLNoSp dst, iRegL src1, vReg src2, vRegMask_V0 v0
   predicate(Matcher::vector_element_basic_type(n->in(2)) == T_LONG);
   match(Set dst (XorReductionV (Binary src1 src2) v0));
   effect(TEMP tmp);
-  ins_riscv_vset(false);
+  ins_riscv_vset(riscv_vset_from_node(this, req, Assembler::m1, Assembler::mu));
   format %{ "reduce_xorL_masked $dst, $src1, $src2, $v0\t# KILL $tmp" %}
   ins_encode %{
     BasicType bt = Matcher::vector_element_basic_type(this, $src2);
     __ reduce_integral_v($dst$$Register, $src1$$Register,
                          as_VectorRegister($src2$$reg), as_VectorRegister($tmp$$reg),
                          this->ideal_Opcode(), bt, Matcher::vector_length(this, $src2),
-                         Assembler::v0_t);
+                         Assembler::v0_t, false /* set_vtype */);
   %}
   ins_pipe(pipe_slow);
 %}
@@ -2441,13 +2457,14 @@ instruct reduce_add(iRegINoSp dst, iRegIorL2I src1, vReg src2, vReg tmp) %{
             Matcher::vector_element_basic_type(n->in(2)) == T_INT);
   match(Set dst (AddReductionVI src1 src2));
   effect(TEMP tmp);
-  ins_riscv_vset(false);
+  ins_riscv_vset(riscv_vset_from_node(this, req, Assembler::m1, Assembler::ma));
   format %{ "reduce_add $dst, $src1, $src2\t# KILL $tmp" %}
   ins_encode %{
     BasicType bt = Matcher::vector_element_basic_type(this, $src2);
     __ reduce_integral_v($dst$$Register, $src1$$Register,
                          as_VectorRegister($src2$$reg), as_VectorRegister($tmp$$reg),
-                         this->ideal_Opcode(), bt, Matcher::vector_length(this, $src2));
+                         this->ideal_Opcode(), bt, Matcher::vector_length(this, $src2),
+                         Assembler::unmasked, false /* set_vtype */);
   %}
   ins_pipe(pipe_slow);
 %}
@@ -2456,13 +2473,14 @@ instruct reduce_addL(iRegLNoSp dst, iRegL src1, vReg src2, vReg tmp) %{
   predicate(Matcher::vector_element_basic_type(n->in(2)) == T_LONG);
   match(Set dst (AddReductionVL src1 src2));
   effect(TEMP tmp);
-  ins_riscv_vset(false);
+  ins_riscv_vset(riscv_vset_from_node(this, req, Assembler::m1, Assembler::ma));
   format %{ "reduce_addL $dst, $src1, $src2\t# KILL $tmp" %}
   ins_encode %{
     BasicType bt = Matcher::vector_element_basic_type(this, $src2);
     __ reduce_integral_v($dst$$Register, $src1$$Register,
                          as_VectorRegister($src2$$reg), as_VectorRegister($tmp$$reg),
-                         this->ideal_Opcode(), bt, Matcher::vector_length(this, $src2));
+                         this->ideal_Opcode(), bt, Matcher::vector_length(this, $src2),
+                         Assembler::unmasked, false /* set_vtype */);
   %}
   ins_pipe(pipe_slow);
 %}
@@ -2479,7 +2497,7 @@ instruct reduce_addF_ordered(fRegF dst, fRegF src1, vReg src2, vReg tmp) %{
   predicate(n->as_Reduction()->requires_strict_order());
   match(Set dst (AddReductionVF src1 src2));
   effect(TEMP tmp);
-  ins_riscv_vset(riscv_vset_from_node(this, req));
+  ins_riscv_vset(riscv_vset_from_node(this, req, Assembler::m1, Assembler::ma));
   format %{ "reduce_addF_ordered $dst, $src1, $src2\t# KILL $tmp" %}
   ins_encode %{
     __ vfmv_s_f(as_VectorRegister($tmp$$reg), $src1$$FloatRegister);
@@ -2494,7 +2512,7 @@ instruct reduce_addF_unordered(fRegF dst, fRegF src1, vReg src2, vReg tmp) %{
   predicate(!n->as_Reduction()->requires_strict_order());
   match(Set dst (AddReductionVF src1 src2));
   effect(TEMP tmp);
-  ins_riscv_vset(riscv_vset_from_node(this, req));
+  ins_riscv_vset(riscv_vset_from_node(this, req, Assembler::m1, Assembler::ma));
   format %{ "reduce_addF_unordered $dst, $src1, $src2\t# KILL $tmp" %}
   ins_encode %{
     __ vfmv_s_f(as_VectorRegister($tmp$$reg), $src1$$FloatRegister);
@@ -2509,7 +2527,7 @@ instruct reduce_addD_ordered(fRegD dst, fRegD src1, vReg src2, vReg tmp) %{
   predicate(n->as_Reduction()->requires_strict_order());
   match(Set dst (AddReductionVD src1 src2));
   effect(TEMP tmp);
-  ins_riscv_vset(riscv_vset_from_node(this, req));
+  ins_riscv_vset(riscv_vset_from_node(this, req, Assembler::m1, Assembler::ma));
   format %{ "reduce_addD_ordered $dst, $src1, $src2\t# KILL $tmp" %}
   ins_encode %{
     __ vfmv_s_f(as_VectorRegister($tmp$$reg), $src1$$FloatRegister);
@@ -2524,7 +2542,7 @@ instruct reduce_addD_unordered(fRegD dst, fRegD src1, vReg src2, vReg tmp) %{
   predicate(!n->as_Reduction()->requires_strict_order());
   match(Set dst (AddReductionVD src1 src2));
   effect(TEMP tmp);
-  ins_riscv_vset(riscv_vset_from_node(this, req));
+  ins_riscv_vset(riscv_vset_from_node(this, req, Assembler::m1, Assembler::ma));
   format %{ "reduce_addD_unordered $dst, $src1, $src2\t# KILL $tmp" %}
   ins_encode %{
     __ vfmv_s_f(as_VectorRegister($tmp$$reg), $src1$$FloatRegister);
@@ -2543,14 +2561,14 @@ instruct reduce_add_masked(iRegINoSp dst, iRegIorL2I src1, vReg src2, vRegMask_V
             Matcher::vector_element_basic_type(n->in(2)) == T_INT);
   match(Set dst (AddReductionVI (Binary src1 src2) v0));
   effect(TEMP tmp);
-  ins_riscv_vset(false);
+  ins_riscv_vset(riscv_vset_from_node(this, req, Assembler::m1, Assembler::mu));
   format %{ "reduce_add_masked $dst, $src1, $src2, $v0\t# KILL $tmp" %}
   ins_encode %{
     BasicType bt = Matcher::vector_element_basic_type(this, $src2);
     __ reduce_integral_v($dst$$Register, $src1$$Register,
                          as_VectorRegister($src2$$reg), as_VectorRegister($tmp$$reg),
                          this->ideal_Opcode(), bt, Matcher::vector_length(this, $src2),
-                         Assembler::v0_t);
+                         Assembler::v0_t, false /* set_vtype */);
   %}
   ins_pipe(pipe_slow);
 %}
@@ -2559,14 +2577,14 @@ instruct reduce_addL_masked(iRegLNoSp dst, iRegL src1, vReg src2, vRegMask_V0 v0
   predicate(Matcher::vector_element_basic_type(n->in(2)) == T_LONG);
   match(Set dst (AddReductionVL (Binary src1 src2) v0));
   effect(TEMP tmp);
-  ins_riscv_vset(false);
+  ins_riscv_vset(riscv_vset_from_node(this, req, Assembler::m1, Assembler::mu));
   format %{ "reduce_addL_masked $dst, $src1, $src2, $v0\t# KILL $tmp" %}
   ins_encode %{
     BasicType bt = Matcher::vector_element_basic_type(this, $src2);
     __ reduce_integral_v($dst$$Register, $src1$$Register,
                          as_VectorRegister($src2$$reg), as_VectorRegister($tmp$$reg),
                          this->ideal_Opcode(), bt, Matcher::vector_length(this, $src2),
-                         Assembler::v0_t);
+                         Assembler::v0_t, false /* set_vtype */);
   %}
   ins_pipe(pipe_slow);
 %}
@@ -2607,13 +2625,14 @@ instruct vreduce_max(iRegINoSp dst, iRegIorL2I src1, vReg src2, vReg tmp) %{
             Matcher::vector_element_basic_type(n->in(2)) == T_INT);
   match(Set dst (MaxReductionV src1 src2));
   effect(TEMP tmp);
-  ins_riscv_vset(false);
+  ins_riscv_vset(riscv_vset_from_node(this, req, Assembler::m1, Assembler::ma));
   format %{ "vreduce_max $dst, $src1, $src2\t# KILL $tmp" %}
   ins_encode %{
     BasicType bt = Matcher::vector_element_basic_type(this, $src2);
     __ reduce_integral_v($dst$$Register, $src1$$Register,
                          as_VectorRegister($src2$$reg), as_VectorRegister($tmp$$reg),
-                         this->ideal_Opcode(), bt, Matcher::vector_length(this, $src2));
+                         this->ideal_Opcode(), bt, Matcher::vector_length(this, $src2),
+                         Assembler::unmasked, false /* set_vtype */);
   %}
   ins_pipe(pipe_slow);
 %}
@@ -2622,13 +2641,14 @@ instruct vreduce_maxL(iRegLNoSp dst, iRegL src1, vReg src2, vReg tmp) %{
   predicate(Matcher::vector_element_basic_type(n->in(2)) == T_LONG);
   match(Set dst (MaxReductionV src1 src2));
   effect(TEMP tmp);
-  ins_riscv_vset(false);
+  ins_riscv_vset(riscv_vset_from_node(this, req, Assembler::m1, Assembler::ma));
   format %{ "vreduce_maxL $dst, $src1, $src2\t# KILL $tmp" %}
   ins_encode %{
     BasicType bt = Matcher::vector_element_basic_type(this, $src2);
     __ reduce_integral_v($dst$$Register, $src1$$Register,
                          as_VectorRegister($src2$$reg), as_VectorRegister($tmp$$reg),
-                         this->ideal_Opcode(), bt, Matcher::vector_length(this, $src2));
+                         this->ideal_Opcode(), bt, Matcher::vector_length(this, $src2),
+                         Assembler::unmasked, false /* set_vtype */);
   %}
   ins_pipe(pipe_slow);
 %}
@@ -2641,14 +2661,14 @@ instruct vreduce_max_masked(iRegINoSp dst, iRegIorL2I src1, vReg src2, vRegMask_
             Matcher::vector_element_basic_type(n->in(2)) == T_INT);
   match(Set dst (MaxReductionV (Binary src1 src2) v0));
   effect(TEMP tmp);
-  ins_riscv_vset(false);
+  ins_riscv_vset(riscv_vset_from_node(this, req, Assembler::m1, Assembler::mu));
   format %{ "vreduce_max_masked $dst, $src1, $src2, $v0\t# KILL $tmp" %}
   ins_encode %{
     BasicType bt = Matcher::vector_element_basic_type(this, $src2);
     __ reduce_integral_v($dst$$Register, $src1$$Register,
                          as_VectorRegister($src2$$reg), as_VectorRegister($tmp$$reg),
                          this->ideal_Opcode(), bt, Matcher::vector_length(this, $src2),
-                         Assembler::v0_t);
+                         Assembler::v0_t, false /* set_vtype */);
   %}
   ins_pipe(pipe_slow);
 %}
@@ -2657,14 +2677,14 @@ instruct vreduce_maxL_masked(iRegLNoSp dst, iRegL src1, vReg src2, vRegMask_V0 v
   predicate(Matcher::vector_element_basic_type(n->in(2)) == T_LONG);
   match(Set dst (MaxReductionV (Binary src1 src2) v0));
   effect(TEMP tmp);
-  ins_riscv_vset(false);
+  ins_riscv_vset(riscv_vset_from_node(this, req, Assembler::m1, Assembler::mu));
   format %{ "vreduce_maxL_masked $dst, $src1, $src2, $v0\t# KILL $tmp" %}
   ins_encode %{
     BasicType bt = Matcher::vector_element_basic_type(this, $src2);
     __ reduce_integral_v($dst$$Register, $src1$$Register,
                          as_VectorRegister($src2$$reg), as_VectorRegister($tmp$$reg),
                          this->ideal_Opcode(), bt, Matcher::vector_length(this, $src2),
-                         Assembler::v0_t);
+                         Assembler::v0_t, false /* set_vtype */);
   %}
   ins_pipe(pipe_slow);
 %}
@@ -2677,13 +2697,14 @@ instruct vreduce_min(iRegINoSp dst, iRegIorL2I src1, vReg src2, vReg tmp) %{
             Matcher::vector_element_basic_type(n->in(2)) == T_INT);
   match(Set dst (MinReductionV src1 src2));
   effect(TEMP tmp);
-  ins_riscv_vset(false);
+  ins_riscv_vset(riscv_vset_from_node(this, req, Assembler::m1, Assembler::ma));
   format %{ "vreduce_min $dst, $src1, $src2\t# KILL $tmp" %}
   ins_encode %{
     BasicType bt = Matcher::vector_element_basic_type(this, $src2);
     __ reduce_integral_v($dst$$Register, $src1$$Register,
                          as_VectorRegister($src2$$reg), as_VectorRegister($tmp$$reg),
-                         this->ideal_Opcode(), bt, Matcher::vector_length(this, $src2));
+                         this->ideal_Opcode(), bt, Matcher::vector_length(this, $src2),
+                         Assembler::unmasked, false /* set_vtype */);
   %}
   ins_pipe(pipe_slow);
 %}
@@ -2692,13 +2713,14 @@ instruct vreduce_minL(iRegLNoSp dst, iRegL src1, vReg src2, vReg tmp) %{
   predicate(Matcher::vector_element_basic_type(n->in(2)) == T_LONG);
   match(Set dst (MinReductionV src1 src2));
   effect(TEMP tmp);
-  ins_riscv_vset(false);
+  ins_riscv_vset(riscv_vset_from_node(this, req, Assembler::m1, Assembler::ma));
   format %{ "vreduce_minL $dst, $src1, $src2\t# KILL $tmp" %}
   ins_encode %{
     BasicType bt = Matcher::vector_element_basic_type(this, $src2);
     __ reduce_integral_v($dst$$Register, $src1$$Register,
                          as_VectorRegister($src2$$reg), as_VectorRegister($tmp$$reg),
-                         this->ideal_Opcode(), bt, Matcher::vector_length(this, $src2));
+                         this->ideal_Opcode(), bt, Matcher::vector_length(this, $src2),
+                         Assembler::unmasked, false /* set_vtype */);
   %}
   ins_pipe(pipe_slow);
 %}
@@ -2711,14 +2733,14 @@ instruct vreduce_min_masked(iRegINoSp dst, iRegIorL2I src1, vReg src2, vRegMask_
             Matcher::vector_element_basic_type(n->in(2)) == T_INT);
   match(Set dst (MinReductionV (Binary src1 src2) v0));
   effect(TEMP tmp);
-  ins_riscv_vset(false);
+  ins_riscv_vset(riscv_vset_from_node(this, req, Assembler::m1, Assembler::mu));
   format %{ "vreduce_min_masked $dst, $src1, $src2, $v0\t# KILL $tmp" %}
   ins_encode %{
     BasicType bt = Matcher::vector_element_basic_type(this, $src2);
     __ reduce_integral_v($dst$$Register, $src1$$Register,
                          as_VectorRegister($src2$$reg), as_VectorRegister($tmp$$reg),
                          this->ideal_Opcode(), bt, Matcher::vector_length(this, $src2),
-                         Assembler::v0_t);
+                         Assembler::v0_t, false /* set_vtype */);
   %}
   ins_pipe(pipe_slow);
 %}
@@ -2727,14 +2749,14 @@ instruct vreduce_minL_masked(iRegLNoSp dst, iRegL src1, vReg src2, vRegMask_V0 v
   predicate(Matcher::vector_element_basic_type(n->in(2)) == T_LONG);
   match(Set dst (MinReductionV (Binary src1 src2) v0));
   effect(TEMP tmp);
-  ins_riscv_vset(false);
+  ins_riscv_vset(riscv_vset_from_node(this, req, Assembler::m1, Assembler::mu));
   format %{ "vreduce_minL_masked $dst, $src1, $src2, $v0\t# KILL $tmp" %}
   ins_encode %{
     BasicType bt = Matcher::vector_element_basic_type(this, $src2);
     __ reduce_integral_v($dst$$Register, $src1$$Register,
                          as_VectorRegister($src2$$reg), as_VectorRegister($tmp$$reg),
                          this->ideal_Opcode(), bt, Matcher::vector_length(this, $src2),
-                         Assembler::v0_t);
+                         Assembler::v0_t, false /* set_vtype */);
   %}
   ins_pipe(pipe_slow);
 %}
@@ -2745,13 +2767,14 @@ instruct vreduce_maxF(fRegF dst, fRegF src1, vReg src2, vReg tmp1, vReg tmp2) %{
   predicate(Matcher::vector_element_basic_type(n->in(2)) == T_FLOAT);
   match(Set dst (MaxReductionV src1 src2));
   effect(TEMP_DEF dst, TEMP tmp1, TEMP tmp2);
-  ins_riscv_vset(false);
+  ins_riscv_vset(riscv_vset_from_node(this, req, Assembler::m1, Assembler::ma));
   format %{ "vreduce_maxF $dst, $src1, $src2, $tmp1, $tmp2" %}
   ins_encode %{
     __ reduce_minmax_fp_v($dst$$FloatRegister,
                           $src1$$FloatRegister, as_VectorRegister($src2$$reg),
                           as_VectorRegister($tmp1$$reg), as_VectorRegister($tmp2$$reg),
-                          false /* is_double */, false /* is_min */, Matcher::vector_length(this, $src2));
+                          false /* is_double */, false /* is_min */, Matcher::vector_length(this, $src2),
+                          Assembler::unmasked, false /* set_vtype */);
   %}
   ins_pipe(pipe_slow);
 %}
@@ -2760,13 +2783,14 @@ instruct vreduce_maxD(fRegD dst, fRegD src1, vReg src2, vReg tmp1, vReg tmp2) %{
   predicate(Matcher::vector_element_basic_type(n->in(2)) == T_DOUBLE);
   match(Set dst (MaxReductionV src1 src2));
   effect(TEMP_DEF dst, TEMP tmp1, TEMP tmp2);
-  ins_riscv_vset(false);
+  ins_riscv_vset(riscv_vset_from_node(this, req, Assembler::m1, Assembler::ma));
   format %{ "vreduce_maxD $dst, $src1, $src2, $tmp1, $tmp2" %}
   ins_encode %{
     __ reduce_minmax_fp_v($dst$$FloatRegister,
                           $src1$$FloatRegister, as_VectorRegister($src2$$reg),
                           as_VectorRegister($tmp1$$reg), as_VectorRegister($tmp2$$reg),
-                          true /* is_double */, false /* is_min */, Matcher::vector_length(this, $src2));
+                          true /* is_double */, false /* is_min */, Matcher::vector_length(this, $src2),
+                          Assembler::unmasked, false /* set_vtype */);
   %}
   ins_pipe(pipe_slow);
 %}
@@ -2777,14 +2801,14 @@ instruct vreduce_maxF_masked(fRegF dst, fRegF src1, vReg src2, vRegMask_V0 v0, v
   predicate(Matcher::vector_element_basic_type(n->in(2)) == T_FLOAT);
   match(Set dst (MaxReductionV (Binary src1 src2) v0));
   effect(TEMP_DEF dst, TEMP tmp1, TEMP tmp2);
-  ins_riscv_vset(false);
+  ins_riscv_vset(riscv_vset_from_node(this, req, Assembler::m1, Assembler::mu));
   format %{ "vreduce_maxF_masked $dst, $src1, $src2, $v0\t# KILL $tmp1, $tmp2" %}
   ins_encode %{
     __ reduce_minmax_fp_v($dst$$FloatRegister,
                           $src1$$FloatRegister, as_VectorRegister($src2$$reg),
                           as_VectorRegister($tmp1$$reg), as_VectorRegister($tmp2$$reg),
                           false /* is_double */, false /* is_min */,
-                          Matcher::vector_length(this, $src2), Assembler::v0_t);
+                          Matcher::vector_length(this, $src2), Assembler::v0_t, false /* set_vtype */);
   %}
   ins_pipe(pipe_slow);
 %}
@@ -2793,14 +2817,14 @@ instruct vreduce_maxD_masked(fRegD dst, fRegD src1, vReg src2, vRegMask_V0 v0, v
   predicate(Matcher::vector_element_basic_type(n->in(2)) == T_DOUBLE);
   match(Set dst (MaxReductionV (Binary src1 src2) v0));
   effect(TEMP_DEF dst, TEMP tmp1, TEMP tmp2);
-  ins_riscv_vset(false);
+  ins_riscv_vset(riscv_vset_from_node(this, req, Assembler::m1, Assembler::mu));
   format %{ "vreduce_maxD_masked $dst, $src1, $src2, $v0\t# KILL $tmp1, $tmp2" %}
   ins_encode %{
     __ reduce_minmax_fp_v($dst$$FloatRegister,
                           $src1$$FloatRegister, as_VectorRegister($src2$$reg),
                           as_VectorRegister($tmp1$$reg), as_VectorRegister($tmp2$$reg),
                           true /* is_double */, false /* is_min */,
-                          Matcher::vector_length(this, $src2), Assembler::v0_t);
+                          Matcher::vector_length(this, $src2), Assembler::v0_t, false /* set_vtype */);
   %}
   ins_pipe(pipe_slow);
 %}
@@ -2811,13 +2835,14 @@ instruct vreduce_minF(fRegF dst, fRegF src1, vReg src2, vReg tmp1, vReg tmp2) %{
   predicate(Matcher::vector_element_basic_type(n->in(2)) == T_FLOAT);
   match(Set dst (MinReductionV src1 src2));
   effect(TEMP_DEF dst, TEMP tmp1, TEMP tmp2);
-  ins_riscv_vset(false);
+  ins_riscv_vset(riscv_vset_from_node(this, req, Assembler::m1, Assembler::ma));
   format %{ "vreduce_minF $dst, $src1, $src2, $tmp1, $tmp2" %}
   ins_encode %{
     __ reduce_minmax_fp_v($dst$$FloatRegister,
                           $src1$$FloatRegister, as_VectorRegister($src2$$reg),
                           as_VectorRegister($tmp1$$reg), as_VectorRegister($tmp2$$reg),
-                          false /* is_double */, true /* is_min */, Matcher::vector_length(this, $src2));
+                          false /* is_double */, true /* is_min */, Matcher::vector_length(this, $src2),
+                          Assembler::unmasked, false /* set_vtype */);
   %}
   ins_pipe(pipe_slow);
 %}
@@ -2826,13 +2851,14 @@ instruct vreduce_minD(fRegD dst, fRegD src1, vReg src2, vReg tmp1, vReg tmp2) %{
   predicate(Matcher::vector_element_basic_type(n->in(2)) == T_DOUBLE);
   match(Set dst (MinReductionV src1 src2));
   effect(TEMP_DEF dst, TEMP tmp1, TEMP tmp2);
-  ins_riscv_vset(false);
+  ins_riscv_vset(riscv_vset_from_node(this, req, Assembler::m1, Assembler::ma));
   format %{ "vreduce_minD $dst, $src1, $src2, $tmp1, $tmp2" %}
   ins_encode %{
     __ reduce_minmax_fp_v($dst$$FloatRegister,
                           $src1$$FloatRegister, as_VectorRegister($src2$$reg),
                           as_VectorRegister($tmp1$$reg), as_VectorRegister($tmp2$$reg),
-                          true /* is_double */, true /* is_min */, Matcher::vector_length(this, $src2));
+                          true /* is_double */, true /* is_min */, Matcher::vector_length(this, $src2),
+                          Assembler::unmasked, false /* set_vtype */);
   %}
   ins_pipe(pipe_slow);
 %}
@@ -2843,14 +2869,14 @@ instruct vreduce_minF_masked(fRegF dst, fRegF src1, vReg src2, vRegMask_V0 v0, v
   predicate(Matcher::vector_element_basic_type(n->in(2)) == T_FLOAT);
   match(Set dst (MinReductionV (Binary src1 src2) v0));
   effect(TEMP_DEF dst, TEMP tmp1, TEMP tmp2);
-  ins_riscv_vset(false);
+  ins_riscv_vset(riscv_vset_from_node(this, req, Assembler::m1, Assembler::mu));
   format %{ "vreduce_minF_masked $dst, $src1, $src2, $v0\t# KILL $tmp1, $tmp2" %}
   ins_encode %{
     __ reduce_minmax_fp_v($dst$$FloatRegister,
                           $src1$$FloatRegister, as_VectorRegister($src2$$reg),
                           as_VectorRegister($tmp1$$reg), as_VectorRegister($tmp2$$reg),
                           false /* is_double */, true /* is_min */,
-                          Matcher::vector_length(this, $src2), Assembler::v0_t);
+                          Matcher::vector_length(this, $src2), Assembler::v0_t, false /* set_vtype */);
   %}
   ins_pipe(pipe_slow);
 %}
@@ -2859,14 +2885,14 @@ instruct vreduce_minD_masked(fRegD dst, fRegD src1, vReg src2, vRegMask_V0 v0, v
   predicate(Matcher::vector_element_basic_type(n->in(2)) == T_DOUBLE);
   match(Set dst (MinReductionV (Binary src1 src2) v0));
   effect(TEMP_DEF dst, TEMP tmp1, TEMP tmp2);
-  ins_riscv_vset(false);
+  ins_riscv_vset(riscv_vset_from_node(this, req, Assembler::m1, Assembler::mu));
   format %{ "vreduce_minD_masked $dst, $src1, $src2, $v0\t# KILL $tmp1, $tmp2" %}
   ins_encode %{
     __ reduce_minmax_fp_v($dst$$FloatRegister,
                           $src1$$FloatRegister, as_VectorRegister($src2$$reg),
                           as_VectorRegister($tmp1$$reg), as_VectorRegister($tmp2$$reg),
                           true /* is_double */, true /* is_min */,
-                          Matcher::vector_length(this, $src2), Assembler::v0_t);
+                          Matcher::vector_length(this, $src2), Assembler::v0_t, false /* set_vtype */);
   %}
   ins_pipe(pipe_slow);
 %}
@@ -4544,19 +4570,22 @@ instruct vblend(vReg dst, vReg src1, vReg src2, vRegMask_V0 v0) %{
 instruct vcvtBtoX(vReg dst, vReg src) %{
   match(Set dst (VectorCastB2X src));
   effect(TEMP_DEF dst);
-  ins_riscv_vset(false);
+  ins_riscv_vset(riscv_vset_fixed(is_floating_point_type(Matcher::vector_element_basic_type(this)) ?
+                                  (Matcher::vector_element_basic_type(this) == T_FLOAT ? T_INT : T_LONG) :
+                                  Matcher::vector_element_basic_type(this),
+                                  Matcher::vector_length(this), req, Assembler::m1, Assembler::ma));
   format %{ "vcvtBtoX $dst, $src" %}
   ins_encode %{
     BasicType bt = Matcher::vector_element_basic_type(this);
     if (is_floating_point_type(bt)) {
       __ integer_extend_v(as_VectorRegister($dst$$reg), bt == T_FLOAT ? T_INT : T_LONG,
                           Matcher::vector_length(this), as_VectorRegister($src$$reg), T_BYTE,
-                          true /* is_signed */);
+                          true /* is_signed */, false /* set_vtype */);
       __ vfcvt_f_x_v(as_VectorRegister($dst$$reg), as_VectorRegister($dst$$reg));
     } else {
       __ integer_extend_v(as_VectorRegister($dst$$reg), bt,
                           Matcher::vector_length(this), as_VectorRegister($src$$reg), T_BYTE,
-                          true /* is_signed */);
+                          true /* is_signed */, false /* set_vtype */);
     }
   %}
   ins_pipe(pipe_slow);
@@ -4568,13 +4597,13 @@ instruct vcvtUBtoX(vReg dst, vReg src) %{
             Matcher::vector_element_basic_type(n) == T_LONG);
   match(Set dst (VectorUCastB2X src));
   effect(TEMP_DEF dst);
-  ins_riscv_vset(false);
+  ins_riscv_vset(riscv_vset_from_vect(this, req, Assembler::m1, Assembler::ma));
   format %{ "vcvtUBtoX $dst, $src" %}
   ins_encode %{
     BasicType bt = Matcher::vector_element_basic_type(this);
     __ integer_extend_v(as_VectorRegister($dst$$reg), bt,
                         Matcher::vector_length(this), as_VectorRegister($src$$reg), T_BYTE,
-                        false /* is_signed */);
+                        false /* is_signed */, false /* set_vtype */);
   %}
   ins_pipe(pipe_slow);
 %}
@@ -4598,12 +4627,12 @@ instruct vcvtStoX(vReg dst, vReg src) %{
             Matcher::vector_element_basic_type(n) == T_LONG);
   match(Set dst (VectorCastS2X src));
   effect(TEMP_DEF dst);
-  ins_riscv_vset(false);
+  ins_riscv_vset(riscv_vset_from_vect(this, req, Assembler::m1, Assembler::ma));
   format %{ "vcvtStoX $dst, $src" %}
   ins_encode %{
     __ integer_extend_v(as_VectorRegister($dst$$reg), Matcher::vector_element_basic_type(this),
                         Matcher::vector_length(this), as_VectorRegister($src$$reg), T_SHORT,
-                        true /* is_signed */);
+                        true /* is_signed */, false /* set_vtype */);
   %}
   ins_pipe(pipe_slow);
 %}
@@ -4632,12 +4661,12 @@ instruct vcvtUStoX(vReg dst, vReg src) %{
             Matcher::vector_element_basic_type(n) == T_LONG);
   match(Set dst (VectorUCastS2X src));
   effect(TEMP_DEF dst);
-  ins_riscv_vset(false);
+  ins_riscv_vset(riscv_vset_from_vect(this, req, Assembler::m1, Assembler::ma));
   format %{ "vcvtUStoX $dst, $src" %}
   ins_encode %{
     __ integer_extend_v(as_VectorRegister($dst$$reg), Matcher::vector_element_basic_type(this),
                         Matcher::vector_length(this), as_VectorRegister($src$$reg), T_SHORT,
-                        false /* is_signed */);
+                        false /* is_signed */, false /* set_vtype */);
   %}
   ins_pipe(pipe_slow);
 %}
@@ -4662,12 +4691,12 @@ instruct vcvtItoL(vReg dst, vReg src) %{
   predicate(Matcher::vector_element_basic_type(n) == T_LONG);
   match(Set dst (VectorCastI2X src));
   effect(TEMP_DEF dst);
-  ins_riscv_vset(false);
+  ins_riscv_vset(riscv_vset_fixed(T_LONG, Matcher::vector_length(this), req, Assembler::m1, Assembler::ma));
   format %{ "vcvtItoL $dst, $src" %}
   ins_encode %{
     __ integer_extend_v(as_VectorRegister($dst$$reg), T_LONG,
                         Matcher::vector_length(this), as_VectorRegister($src$$reg), T_INT,
-                        true /* is_signed */);
+                        true /* is_signed */, false /* set_vtype */);
   %}
   ins_pipe(pipe_slow);
 %}
@@ -4676,12 +4705,12 @@ instruct vcvtUItoL(vReg dst, vReg src) %{
   predicate(Matcher::vector_element_basic_type(n) == T_LONG);
   match(Set dst (VectorUCastI2X src));
   effect(TEMP_DEF dst);
-  ins_riscv_vset(false);
+  ins_riscv_vset(riscv_vset_fixed(T_LONG, Matcher::vector_length(this), req, Assembler::m1, Assembler::ma));
   format %{ "vcvtUItoL $dst, $src" %}
   ins_encode %{
     __ integer_extend_v(as_VectorRegister($dst$$reg), T_LONG,
                         Matcher::vector_length(this), as_VectorRegister($src$$reg), T_INT,
-                        false /* is_signed */);
+                        false /* is_signed */, false /* set_vtype */);
   %}
   ins_pipe(pipe_slow);
 %}
@@ -5087,12 +5116,12 @@ instruct vsignum_reg(vReg dst, vReg zero, vReg one, vRegMask_V0 v0) %{
   match(Set dst (SignumVF dst (Binary zero one)));
   match(Set dst (SignumVD dst (Binary zero one)));
   effect(TEMP_DEF dst, TEMP v0);
-  ins_riscv_vset(false);
+  ins_riscv_vset(riscv_vset_from_vect(this, req, Assembler::m1, Assembler::mu));
   format %{ "vsignum $dst, $dst\t" %}
   ins_encode %{
     BasicType bt = Matcher::vector_element_basic_type(this);
     __ signum_fp_v(as_VectorRegister($dst$$reg), as_VectorRegister($one$$reg),
-                   bt, Matcher::vector_length(this));
+                   bt, Matcher::vector_length(this), false /* set_vtype */);
   %}
   ins_pipe(pipe_slow);
 %}
@@ -5102,13 +5131,14 @@ instruct vsignum_reg(vReg dst, vReg zero, vReg one, vRegMask_V0 v0) %{
 instruct vround_f(vReg dst, vReg src, fRegF tmp, vRegMask_V0 v0) %{
   match(Set dst (RoundVF src));
   effect(TEMP_DEF dst, TEMP tmp, TEMP v0);
-  ins_riscv_vset(false);
+  ins_riscv_vset(riscv_vset_from_vect(this, req, Assembler::m1, Assembler::mu));
   format %{ "java_round_float_v $dst, $src\t" %}
   ins_encode %{
     BasicType bt = Matcher::vector_element_basic_type(this);
     uint vector_length = Matcher::vector_length(this);
     __ java_round_float_v(as_VectorRegister($dst$$reg), as_VectorRegister($src$$reg),
-                          as_FloatRegister($tmp$$reg), bt, vector_length);
+                          as_FloatRegister($tmp$$reg), bt, vector_length,
+                          false /* set_vtype */);
   %}
   ins_pipe(pipe_slow);
 %}
@@ -5116,13 +5146,14 @@ instruct vround_f(vReg dst, vReg src, fRegF tmp, vRegMask_V0 v0) %{
 instruct vround_d(vReg dst, vReg src, fRegD tmp, vRegMask_V0 v0) %{
   match(Set dst (RoundVD src));
   effect(TEMP_DEF dst, TEMP tmp, TEMP v0);
-  ins_riscv_vset(false);
+  ins_riscv_vset(riscv_vset_from_vect(this, req, Assembler::m1, Assembler::mu));
   format %{ "java_round_double_v $dst, $src\t" %}
   ins_encode %{
     BasicType bt = Matcher::vector_element_basic_type(this);
     uint vector_length = Matcher::vector_length(this);
     __ java_round_double_v(as_VectorRegister($dst$$reg), as_VectorRegister($src$$reg),
-                           as_FloatRegister($tmp$$reg), bt, vector_length);
+                           as_FloatRegister($tmp$$reg), bt, vector_length,
+                           false /* set_vtype */);
   %}
   ins_pipe(pipe_slow);
 %}
